File: reference/src/tq_reference/philox.py
"""RNG-001…005 reference for the shared Philox4x32-10 test stream.

Contract: ``reference/contracts/philox.md``. This is test infrastructure, not the production
token sampler. The generator is a counter-based bijection with no state, so the same counter and
key always give the same block and a stream is only an assignment of counters.

One round multiplies ``M0 * c0`` and ``M1 * c2`` as full 64-bit products, splits each into a high
and a low half, and emits ``(hi1 ^ c1 ^ k0, lo1, hi0 ^ c3 ^ k1, lo0)``. Round ``r`` uses the key
``(k0 + r*W0, k1 + r*W1)`` modulo ``2**32``, so round zero uses the key exactly as supplied.

The word at logical offset ``i`` of a stream is lane ``i % 4`` of the block whose key is the seed
split low word first, and whose counter is ``[(i//4) low, (i//4) high, stream low, stream high]``.
Each stream owns ``2**64`` words and never carries into another stream's domain.
"""

import numpy as np

MASK32 = (1 << 32) - 1
MAX_U64 = (1 << 64) - 1
M0 = 0xD2511F53
M1 = 0xCD9E8D57
W0 = 0x9E3779B9
W1 = 0xBB67AE85
ROUNDS = 10
UNIFORM_SCALE = np.float32(1.0 / 16777216.0)  # 2**-24, exact in float32
UINT32 = np.dtype(np.uint32)


class PhiloxError(ValueError):
    """Invalid input to the RNG-001…005 reference."""


def _require_array(name: str, value, shape: tuple | None = None) -> np.ndarray:
    """Check one array argument. A native-order uint32 dtype is required, never converted."""
    if not isinstance(value, np.ndarray):
        raise PhiloxError(f"{name} must be a numpy.ndarray, got {type(value).__name__}")
    if value.dtype != UINT32:
        raise PhiloxError(f"{name} must have dtype uint32, got {value.dtype}")
    if shape is not None and value.shape != shape:
        raise PhiloxError(f"{name} must have shape {shape}, got {value.shape}")
    return value


def _require_scalar(name: str, value) -> int:
    """Check one unsigned 64-bit scalar. The value is never formatted, so huge ints are safe."""
    if isinstance(value, bool) or not isinstance(value, (int, np.integer)):
        raise PhiloxError(f"{name} must be an integer, got {type(value).__name__}")
    number = int(value)
    if not 0 <= number <= MAX_U64:
        raise PhiloxError(f"{name} must be in [0, 2**64)")
    return number


def _rounds(counter: list[np.ndarray], key0: int, key1: int) -> list[np.ndarray]:
    """Ten rounds over uint32 counter-word arrays; the keys stay Python ints."""
    c0, c1, c2, c3 = counter
    for round_index in range(ROUNDS):
        product0 = c0.astype(np.uint64) * np.uint64(M0)
        product1 = c2.astype(np.uint64) * np.uint64(M1)
        high0 = (product0 >> np.uint64(32)).astype(np.uint32)
        high1 = (product1 >> np.uint64(32)).astype(np.uint32)
        round0 = np.uint32((key0 + round_index * W0) & MASK32)
        round1 = np.uint32((key1 + round_index * W1) & MASK32)
        c0, c1, c2, c3 = (
            high1 ^ c1 ^ round0,
            product1.astype(np.uint32),
            high0 ^ c3 ^ round1,
            product0.astype(np.uint32),
        )
    return [c0, c1, c2, c3]


def block(counter, key) -> np.ndarray:
    """Return the four output words of one Philox4x32-10 block as a fresh uint32 ``[4]`` array."""
    counter = _require_array("counter", counter, (4,))
    key = _require_array("key", key, (2,))
    words = _rounds([counter[i : i + 1] for i in range(4)], int(key[0]), int(key[1]))
    return np.concatenate(words).astype(np.uint32, copy=False)


def words(seed, stream, start, count) -> np.ndarray:
    """Return ``count`` stream words from logical offset ``start`` as a fresh uint32 array."""
    seed = _require_scalar("seed", seed)
    stream = _require_scalar("stream", stream)
    start = _require_scalar("start", start)
    count = _require_scalar("count", count)
    if count and start + count - 1 > MAX_U64:
        raise PhiloxError("word range exceeds u64")
    if not count:
        return np.zeros(0, dtype=np.uint32)

    first = start // 4
    blocks = (start + count - 1) // 4 - first + 1
    group = np.uint64(first) + np.arange(blocks, dtype=np.uint64)
    counter = [
        (group & np.uint64(MASK32)).astype(np.uint32),
        (group >> np.uint64(32)).astype(np.uint32),
        np.full(blocks, stream & MASK32, dtype=np.uint32),
        np.full(blocks, stream >> 32, dtype=np.uint32),
    ]
    lanes = np.stack(_rounds(counter, seed & MASK32, seed >> 32), axis=-1).reshape(-1)
    offset = start - first * 4
    return np.ascontiguousarray(lanes[offset : offset + count])


def uniform_f32(words) -> np.ndarray:
    """Map words onto ``[0, 1)`` as ``(word >> 8) * 2**-24``, a grid of ``2**24`` float32 values.

    The top 24 bits are exactly representable in float32 and the scale is a power of two, so the
    product is exact. The largest result is ``1 - 2**-24``, never ``1``. The result keeps the
    input's shape, including a zero-dimensional array for scalar-shaped input.
    """
    values = _require_array("words", words)
    scaled = np.right_shift(values, np.uint32(8)).astype(np.float32) * UNIFORM_SCALE
    return np.asarray(scaled, dtype=np.float32, order="C")

File: crates/tq-testkit/src/philox.rs
//! Philox4x32-10 and the `tq-philox-v1` stream mapping.
//!
//! Contract: `reference/contracts/philox.md`. The generator is a counter-based bijection, so it
//! keeps no state: the same `(counter, key)` always gives the same block, and a stream is only an
//! assignment of counters. Written independently from the published algorithm.
//!
//! One round multiplies `M0 * c0` and `M1 * c2` as full 64-bit products, splits each into a high
//! and a low half, and emits `(hi1 ^ c1 ^ k0, lo1, hi0 ^ c3 ^ k1, lo0)`. Round `r` uses the key
//! `(k0 + r*W0, k1 + r*W1)` modulo `2^32`, so round zero uses the key exactly as supplied.
//!
//! The word at logical offset `i` of a stream is lane `i % 4` of the block whose key is the seed
//! split low word first, and whose counter is `[(i/4) low, (i/4) high, stream low, stream high]`.
//! Every stream therefore owns `2^64` words and never carries into another stream's domain.

use core::error::Error;
use core::fmt;

/// Multiplier applied to counter word 0.
const M0: u64 = 0xd251_1f53;
/// Multiplier applied to counter word 2.
const M1: u64 = 0xcd9e_8d57;
/// Key increment for word 0, the golden ratio.
const W0: u32 = 0x9e37_79b9;
/// Key increment for word 1, `sqrt(3) - 1`.
const W1: u32 = 0xbb67_ae85;
/// Rounds in Philox4x32-10.
const ROUNDS: usize = 10;
/// `2^-24`, exact in `f32` because both the numerator and the denominator are powers of two.
const UNIFORM_SCALE: f32 = 1.0 / 16_777_216.0;

/// This slice supports 32- and 64-bit pointer widths, so a slice length always fits in `u64`.
const _: () = assert!(usize::BITS <= u64::BITS);

/// The requested words do not all fit in the `2^64` addressable words of a stream.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct RangeError;

impl fmt::Display for RangeError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str("word range exceeds u64")
    }
}

impl Error for RangeError {}

/// Splits the full product of a 32-bit value and a multiplier into its high and low halves.
#[expect(
    clippy::cast_possible_truncation,
    reason = "The shift and the implicit mask each select exactly one 32-bit half of the product"
)]
const fn halves(multiplier: u64, value: u32) -> (u32, u32) {
    let product = multiplier * value as u64;
    ((product >> 32) as u32, product as u32)
}

/// Returns the four output words of one Philox4x32-10 block.
#[must_use]
pub fn block(counter: [u32; 4], key: [u32; 2]) -> [u32; 4] {
    let [mut c0, mut c1, mut c2, mut c3] = counter;
    let [mut k0, mut k1] = key;
    for _ in 0..ROUNDS {
        let (hi0, lo0) = halves(M0, c0);
        let (hi1, lo1) = halves(M1, c2);
        (c0, c1, c2, c3) = (hi1 ^ c1 ^ k0, lo1, hi0 ^ c3 ^ k1, lo0);
        // The final bump is never used; keeping it unconditional keeps the round body uniform.
        k0 = k0.wrapping_add(W0);
        k1 = k1.wrapping_add(W1);
    }
    [c0, c1, c2, c3]
}

/// Writes the stream's words at logical offsets `start ..= start + out.len() - 1`.
///
/// # Errors
///
/// Returns [`RangeError`] if the last requested offset would leave `u64`. The buffer is then
/// left exactly as the caller passed it; nothing is wrapped and nothing is partially filled.
pub fn fill_words(seed: u64, stream: u64, start: u64, out: &mut [u32]) -> Result<(), RangeError> {
    let Some(last) = out.len().checked_sub(1) else {
        return Ok(()); // An empty request is valid at every start, including u64::MAX.
    };
    start.checked_add(length_as_u64(last)).ok_or(RangeError)?;

    let key = [low(seed), high(seed)];
    let mut group = start / 4;
    let mut lane = lane_of(start);
    let mut rest = out;
    while !rest.is_empty() {
        let words = block([low(group), high(group), low(stream), high(stream)], key);
        let taken = (4 - lane).min(rest.len());
        rest[..taken].copy_from_slice(&words[lane..lane + taken]);
        rest = &mut rest[taken..];
        // Only the first block can start mid-way, and the validated range keeps group below 2^62.
        lane = 0;
        group += 1;
    }
    Ok(())
}

/// Maps a word onto `[0, 1)` as `(word >> 8) * 2^-24`, a grid of `2^24` float32 values.
///
/// The top 24 bits are exactly representable in float32 and the scale is a power of two, so the
/// product is exact. The largest result is `1 - 2^-24`, never `1`.
#[must_use]
#[expect(
    clippy::cast_precision_loss,
    reason = "The shifted word is below 2^24, so every value converts to f32 exactly"
)]
pub fn uniform_f32(word: u32) -> f32 {
    (word >> 8) as f32 * UNIFORM_SCALE
}

/// The lane a logical word offset occupies inside its block, always `0..=3`.
const fn lane_of(index: u64) -> usize {
    (index % 4) as usize
}

/// Widens a slice length, lossless on every supported pointer width per the assertion above.
const fn length_as_u64(length: usize) -> u64 {
    length as u64
}

/// The low 32 bits of a 64-bit value.
#[expect(
    clippy::cast_possible_truncation,
    reason = "Truncation to the low half is the intended split"
)]
const fn low(value: u64) -> u32 {
    value as u32
}

/// The high 32 bits of a 64-bit value.
const fn high(value: u64) -> u32 {
    (value >> 32) as u32
}
